### PR TITLE
Change how compiler-builtins gets many CGUs

### LINKED ISSUE #73135: compiler-builtins: Int trait functions are not inlined on wasm
I'm currently doing some benchmarks and I've seen that `__multi3` is taking a really long time to calculate on wasm. Turns out that none of the helper functions it uses are inlined at all, so it does a ton of unnecessary calls and misses a lot of potential optimizations:

```wasm
(func $compiler_builtins::int::mul::__multi3::h134a3ac9dcea74e0 (type $t39) (param $p0 i32) (param $p1 i64) (param $p2 i64) (param $p3 i64) (param $p4 i64)
    (local $l5 i32) (local $l6 i64) (local $l7 i64) (local $l8 i64)
    global.get $g0
    i32.const 16
    i32.sub
    local.tee $l5
    global.set $g0
    local.get $p1
    local.get $p2
    call $<i128_as_compiler_builtins::int::LargeInt>::low::h654cf3febd0eecf8
    i64.const 4294967295
    i64.and
    local.get $p3
    local.get $p4
    call $<i128_as_compiler_builtins::int::LargeInt>::low::h654cf3febd0eecf8
    i64.const 4294967295
    i64.and
    call $<i64_as_compiler_builtins::int::Int>::wrapping_mul::h558368f41dd93d09
    local.set $l6
    local.get $p1
    local.get $p2
    call $<i128_as_compiler_builtins::int::LargeInt>::low::h654cf3febd0eecf8
    i64.const 32
    i64.shr_u
    local.get $p3
    local.get $p4
    call $<i128_as_compiler_builtins::int::LargeInt>::low::h654cf3febd0eecf8
    i64.const 4294967295
    i64.and
    call $<i64_as_compiler_builtins::int::Int>::wrapping_mul::h558368f41dd93d09
    local.get $l6
    i64.const 32
    i64.shr_u
    i64.add
    local.tee $l7
    i64.const 32
    i64.shr_u
    call $<i64_as_compiler_builtins::int::Int>::from_unsigned::h7efafe9ef60e0310
    local.set $l8
    local.get $l5
    local.get $p3
    local.get $p4
    call $<i128_as_compiler_builtins::int::LargeInt>::low::h654cf3febd0eecf8
    i64.const 32
    i64.shr_u
    local.get $p1
    local.get $p2
    call $<i128_as_compiler_builtins::int::LargeInt>::low::h654cf3febd0eecf8
    i64.const 4294967295
    i64.and
    call $<i64_as_compiler_builtins::int::Int>::wrapping_mul::h558368f41dd93d09
    local.get $l7
    i64.const 4294967295
    i64.and
    i64.add
    local.tee $l7
    i64.const 32
    i64.shl
    local.get $l6
    i64.const 4294967295
    i64.and
    i64.or
    local.get $l8
    local.get $l7
    i64.const 32
    i64.shr_u
    call $<i64_as_compiler_builtins::int::Int>::from_unsigned::h7efafe9ef60e0310
    i64.add
    local.get $p1
    local.get $p2
    call $<i128_as_compiler_builtins::int::LargeInt>::low::h654cf3febd0eecf8
    i64.const 32
    i64.shr_u
    local.get $p3
    local.get $p4
    call $<i128_as_compiler_builtins::int::LargeInt>::low::h654cf3febd0eecf8
    i64.const 32
    i64.shr_u
    call $<i64_as_compiler_builtins::int::Int>::wrapping_mul::h558368f41dd93d09
    call $<i64_as_compiler_builtins::int::Int>::from_unsigned::h7efafe9ef60e0310
    i64.add
    local.get $p1
    local.get $p2
    call $<i128_as_compiler_builtins::int::LargeInt>::high::hc703bd80007672ce
    local.get $p3
    local.get $p4
    call $<i128_as_compiler_builtins::int::LargeInt>::low::h654cf3febd0eecf8
    call $<i64_as_compiler_builtins::int::Int>::from_unsigned::h7efafe9ef60e0310
    call $<i64_as_compiler_builtins::int::Int>::wrapping_mul::h558368f41dd93d09
    call $<i64_as_compiler_builtins::int::Int>::wrapping_add::h9472fc521f362e13
    local.get $p1
    local.get $p2
    call $<i128_as_compiler_builtins::int::LargeInt>::low::h654cf3febd0eecf8
    call $<i64_as_compiler_builtins::int::Int>::from_unsigned::h7efafe9ef60e0310
    local.get $p3
    local.get $p4
    call $<i128_as_compiler_builtins::int::LargeInt>::high::hc703bd80007672ce
    call $<i64_as_compiler_builtins::int::Int>::wrapping_mul::h558368f41dd93d09
    call $<i64_as_compiler_builtins::int::Int>::wrapping_add::h9472fc521f362e13
    call $<i128_as_compiler_builtins::int::LargeInt>::from_parts::h8ea13ebf67bbf227
    local.get $l5
    i64.load
    local.set $p3
    local.get $p0
    local.get $l5
    i32.const 8
    i32.add
    i64.load
    i64.store offset=8
    local.get $p0
    local.get $p3
    i64.store
    local.get $l5
    i32.const 16
    i32.add
    global.set $g0)
```

The same is probably true for the Float trait.

### PR DESCRIPTION
**Pull request by alexcrichton**
This commit intends to fix an accidental regression from #70846. The
goal of #70846 was to build compiler-builtins with a maximal number of
CGUs to ensure that each module in the source corresponds to an object
file. This high degree of control for compiler-builtins is desirable to
ensure that there's at most one exported symbol per CGU, ideally
enabling compiler-builtins to not conflict with the system libgcc as
often.

In #70846, however, only part of the compiler understands that
compiler-builtins is built with many CGUs. The rest of the compiler
thinks it's building with `sess.codegen_units()`. Notably the
calculation of `sess.lto()` consults `sess.codegen_units()`, which when
there's only one CGU it disables ThinLTO. This means that
compiler-builtins is built without ThinLTO, which is quite harmful to
performance! This is the root of the cause from #73135 where intrinsics
were found to not be inlining trivial functions.

The fix applied in this commit is to remove the special-casing of
compiler-builtins in the compiler. Instead the build system is now
responsible for special-casing compiler-builtins. It doesn't know
exactly how many CGUs will be needed but it passes a large number that
is assumed to be much greater than the number of source-level modules
needed. After reading the various locations in the compiler source, this
seemed like the best solution rather than adding more and more special
casing in the compiler for compiler-builtins.

Closes #73135

### CODE REVIEW

**Comment by ehuss** [comment]
Will build-std need something similar?

**Comment by tmiasko** [comment]
@ehuss the original implementation used attribute specifically for build-std case. It would be nice if this logic was applied there at some point. There is also a similar case of debug assertions which should be disabled for compiler builtins.

**Comment by alexcrichton** [comment]
For build-std I think it sort of depends. The compiler-builtins works regardless of the number of CGUs it is compiled with, it's just that it can work *best* with system libraries that have conflicting symbols if it has one object per symbol. In that sense for many pure-Rust targets using build-std (e.g. anything embedded) the number of CGUs for compiler-builtins doesn't matter. 

In the limit though, yeah, we may have to embed this logic (like debug assertions)

File: src/test/codegen-units/partitioning/compiler-builtins.rs [hunk 1]
```diff
@@ -1,40 +0,0 @@
-// Verifies that during compiler_builtins compilation the codegen units are kept
-// unmerged. Even when only a single codegen unit is requested with -Ccodegen-units=1.
-//
-// compile-flags: -Zprint-mono-items=eager -Ccodegen-units=1
```
Comment (Mark-Simulacrum) [inline src/test/codegen-units/partitioning/compiler-builtins.rs:4]: Is it worth keeping this test with the codegen-units=1000 logic and making sure we're not merging in that case?

**Review by Mark-Simulacrum — COMMENTED** [review]
r=me

File: src/test/codegen-units/partitioning/compiler-builtins.rs (hunk 1 above)
Comment (alexcrichton) [reply src/test/codegen-units/partitioning/compiler-builtins.rs:4]: I suspect not in the sense that there's already likely tests for that, and this is just falling back on existing functionality.

**Comment by glandium** [comment]
There is a regression in 1.44 when building Firefox with cross-language LTO for OSX (and only in that combination), where libxul fails to link with error messages like:
```
Undefined symbols for architecture x86_64:
  "__ZN52_$LT$u128$u20$as$u20$compiler_builtins..int..Int$GT$12extract_sign17h2f47846013b80758E", referenced from:
      __ZN17compiler_builtins5float4conv13__floatuntidf17h0187dacb8fabec21E in libgkrust.a(compiler_builtins-57149daefc485a08.compiler_builtins.4zbdfqik-cgu.80.rcgu.o)
  "__ZN51_$LT$i32$u20$as$u20$compiler_builtins..int..Int$GT$12aborting_div17hb9ce4562c23cc4a0E", referenced from:
      __ZN17compiler_builtins5float3pow9__powisf217h50ee614f1c4d9a6bE in libgkrust.a(compiler_builtins-57149daefc485a08.compiler_builtins.4zbdfqik-cgu.27.rcgu.o)
  "__ZN52_$LT$i128$u20$as$u20$compiler_builtins..int..Int$GT$12extract_sign17h8299efdd5198e4f5E", referenced from:
      __ZN17compiler_builtins5float4conv11__floattidf17h3723b89a83f7291fE in libgkrust.a(compiler_builtins-57149daefc485a08.compiler_builtins.4zbdfqik-cgu.80.rcgu.o)
ld: symbol(s) not found for architecture x86_64

```

This can be reproduced with nightly-2020-04-08 (42abbd887) and not with nightly-2020-04-07 (6dee5f112). Further narrowing down with rustup-install-toolchain-master came out with a range between 39b62533c7f9d0581a6ea9b9fc2cc51f21c3b5b0 and 42abbd8878d3b67238f3611b0587c704ba94f39c, which would seem to point at #70846. I'm hoping this change will fix it, but for some reason, I can't reproduce the problem in the first place with a local build of a revision known to be faulty in nightly, so I can't check whether this fixes it. At this point, I'll wait for the next nightly following the merge of this PR to test whether it fixes it, otherwise, I'll open a new issue.

**Comment by glandium** [comment]
Corollary: if this fixes it, it would be nice to have in a 1.44.1 (or at least in 1.45).

**Comment by Mark-Simulacrum** [comment]
If this (or some other PR) needs stable backports, then the window for 1.44.1 is rapidly closing, would be good to get them stable-nominated today (for either emergency approval tomorrow's compiler meeting or unilateral approval).

**Comment by alexcrichton** [comment]
@Mark-Simulacrum this only affects the performance of compiler-builtins which doesn't affect many high-profile platforms (wasm is the main motivation for this), so I would personally think it's ok to not include this in a point release.

**Comment by glandium** [comment]
I wish I could test this fix...

**Comment by Mark-Simulacrum** [comment]
@bors try

This'll at least get us a try build which might help with testing?

File: src/bootstrap/bin/rustc.rs [hunk 2]
```diff
@@ -125,6 +125,20 @@ fn main() {
...
+        // The goal here is to place every single intrinsic into its own object
+        // file to avoid symbol clashes with the system libgcc if possible. Note
+        // that this number doesn't actually produce this many object files, we
+        // just don't create more than this number of object files.
+        //
+        // It's a bit of a bummer that we have to pass this here, unfortunately.
+        // Ideally this would be specified through an env var to Cargo so Cargo
+        // knows how many CGUs are for this specific crate, but for now
+        // per-crate configuration isn't specifiable in the environment.
+        if crate_name == Some("compiler_builtins") {
```
Comment (bjorn3) [inline src/bootstrap/bin/rustc.rs:139]:
```toml
[profile.release.package.compiler_builtins]
codegen-units = 10000
```

in the workspace `Cargo.toml`?

Comment (alexcrichton) [reply src/bootstrap/bin/rustc.rs]: Oh nice forgot about that, that does indeed work!

**Comment by alexcrichton** [comment]
https://github.com/rust-lang/rust/pull/63338

**Comment by glandium** [comment]
I can confirm the Firefox linking error mentioned in https://github.com/rust-lang/rust/pull/73136#issuecomment-641734206 does not happen anymore with nightly-2020-06-20, which is the first nightly with this fix. I cannot, however, say anything about nightly-2020-06-19, because rustc crashed during compilation (and so did nightly-2020-06-18).

So it would have been nice in 1.44.1, but the second best thing would be to have it in 1.45.

**Comment by Mark-Simulacrum** [comment]
Beta-nominating as this (apparently) fixes a regression -- @glandium would be good to file an issue in the future for easier tracking on our end (in addition to comments like https://github.com/rust-lang/rust/pull/73136#issuecomment-641734206).

**Comment by alexcrichton** [comment]
FWIW this patch as-is uses a relatively new Cargo feature of profile overrides, and if it's backported I'd recommend double-checking that it works before landing.

**Comment by spastorino** [comment]
Adding `T-compiler` so our tools are able to fetch it and include it on the weekly agenda.

**Comment by pnkfelix** [comment]
Discussed in last week's T-compiler [meeting](https://zulip-archive.rust-lang.org/238009tcompilermeetings/59499weeklymeeting2020062554818.html#201974354)

We are inclined to backport the **earlier version** of this PR that doesn't rely on new cargo stuff. See link above for details.

**Pull request merged**